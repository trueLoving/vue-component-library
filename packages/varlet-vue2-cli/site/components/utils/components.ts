import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
