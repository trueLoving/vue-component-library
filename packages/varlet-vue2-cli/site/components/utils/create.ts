import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
