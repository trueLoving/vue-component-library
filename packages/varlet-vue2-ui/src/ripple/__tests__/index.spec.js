import example from '../example'
import Ripple from '..'
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import { delay, mockOffset, trigger, triggerDrag } from '../../utils/jest'

mockOffset()

test('test ripple example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test ripple plugin', () => {
  Vue.use(Ripple)
  expect(Vue.directive('ripple')).toBeTruthy()
})

const Wrapper = {
  directives: { Ripple },
  data: () => ({
    color: 'yellow',
    disabled: false,
  }),
  template: `
    <div class="ripple-container" style="width: 100px; height: 100px" v-ripple="{ color, disabled }"></div>
  `,
}

test('test ripple show & hide', async () => {
  const wrapper = mount(Wrapper, { attachTo: document.body })
  await trigger(wrapper, 'touchstart')
  await delay(60)
  expect(wrapper.find('.var-ripple').exists()).toBeTruthy()

  await trigger(wrapper, 'touchend')
  await delay(1000)
  expect(wrapper.find('.var-ripple').exists()).toBeFalsy()
  wrapper.destroy()
})

test('test ripple update & color', async () => {
  const wrapper = mount(Wrapper, { attachTo: document.body })
  await wrapper.setData({ color: 'green' })

  await trigger(wrapper, 'touchstart')
  await delay(60)
  expect(wrapper.find('.var-ripple').exists()).toBeTruthy()
  expect(wrapper.find('.var-ripple').element.style.backgroundColor).toBe('green')

  wrapper.destroy()
})

test('test ripple disabled', async () => {
  const wrapper = mount(Wrapper, { attachTo: document.body })
  await wrapper.setData({ disabled: true })

  await trigger(wrapper, 'touchstart')
  await delay(60)
  expect(wrapper.find('.var-ripple').exists()).toBeFalsy()

  wrapper.destroy()
})

test('test ripple touchmove', async () => {
  window.ontouchstart = true
  const wrapper = mount(Wrapper, { attachTo: document.body })

  await triggerDrag(wrapper, 0, 20)
  await delay(60)
  expect(wrapper.find('.var-ripple').exists()).toBeFalsy()

  wrapper.destroy()
})
