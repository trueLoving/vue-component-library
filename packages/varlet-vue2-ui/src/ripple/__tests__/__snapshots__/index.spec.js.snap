// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ripple example 1`] = `
"<div class=\\"var-ripple-example\\">
  <div class=\\"app-type\\">基本使用</div>
  <div class=\\"block var-elevation--2\\">点击</div>
  <div class=\\"app-type\\">自定义颜色</div>
  <div class=\\"block var-elevation--2\\">点击</div>
  <div class=\\"app-type\\">禁用状态</div>
  <div class=\\"block var-elevation--2\\">点击</div>
</div>"
`;
